File: sdks/typescript/ki-browser.ts
/**
 * ki-browser TypeScript SDK
 *
 * Handwritten SDK for the ki-browser REST API.
 * Uses native fetch — works in Node.js 18+, Deno, Bun, and modern browsers.
 *
 * @example
 * ```typescript
 * import { KiBrowser } from "./ki-browser";
 *
 * const browser = new KiBrowser("http://localhost:9222");
 * await browser.navigate("https://example.com");
 * const tabs = await browser.listTabs();
 * const screenshot = await browser.screenshot();
 * ```
 */

// ============================================================================
// Response Types
// ============================================================================

/** Health check response from the API server. */
export interface HealthStatus {
  status: string;
  version: string;
  api_enabled: boolean;
}

/** Information about a browser tab. */
export interface TabInfo {
  id: string;
  url: string;
  title: string;
  is_loading: boolean;
  is_active: boolean;
  can_go_back: boolean;
  can_go_forward: boolean;
}

/** Response from listing all tabs. */
export interface TabsResponse {
  tabs: TabInfo[];
  active_tab_id: string | null;
}

/** Response from creating a new tab. */
export interface NewTabResponse {
  tab_id: string;
  url: string;
}

/** Bounding box of a DOM element in pixels. */
export interface BoundingBox {
  x: number;
  y: number;
  width: number;
  height: number;
}

/** Information about a found DOM element. */
export interface ElementInfo {
  found: boolean;
  tag_name?: string;
  text_content?: string;
  attributes?: Record<string, unknown>;
  bounding_box?: BoundingBox;
  is_visible?: boolean;
}

/** Screenshot capture response with base64 encoded image data. */
export interface ScreenshotResponse {
  data: string;
  format: string;
  width: number;
  height: number;
}

/** Response from JavaScript evaluation. */
export interface EvaluateResponse {
  result: unknown;
}

/** API server status. */
export interface ApiStatus {
  enabled: boolean;
  port: number;
  connected_clients: number;
}

// ============================================================================
// Request Types
// ============================================================================

/** Options for creating a new tab. */
export interface NewTabOptions {
  url?: string;
  active?: boolean;
}

/** Options for clicking. */
export interface ClickOptions {
  x?: number;
  y?: number;
  selector?: string;
  button?: "left" | "right" | "middle";
  tab_id?: string;
}

/** Options for typing text. */
export interface TypeOptions {
  selector?: string;
  clear_first?: boolean;
  tab_id?: string;
}

/** Options for scrolling. */
export interface ScrollOptions {
  x?: number;
  y?: number;
  delta_x?: number;
  delta_y?: number;
  selector?: string;
  behavior?: "auto" | "smooth" | "instant";
  tab_id?: string;
}

/** Options for taking a screenshot. */
export interface ScreenshotOptions {
  format?: "png" | "jpeg";
  quality?: number;
  full_page?: boolean;
  selector?: string;
  tab_id?: string;
}

/** Options for finding a DOM element. */
export interface FindElementOptions {
  timeout?: number;
  tab_id?: string;
}

/** Options for evaluating JavaScript. */
export interface EvaluateOptions {
  await_promise?: boolean;
  tab_id?: string;
}

// ============================================================================
// Errors
// ============================================================================

/** Base error for ki-browser SDK operations. */
export class KiBrowserError extends Error {
  public statusCode?: number;

  constructor(message: string, statusCode?: number) {
    super(message);
    this.name = "KiBrowserError";
    this.statusCode = statusCode;
  }
}

/** Raised when the API is disabled (HTTP 503). */
export class ApiDisabledError extends KiBrowserError {
  constructor(message: string = "API is disabled") {
    super(message, 503);
    this.name = "ApiDisabledError";
  }
}

/** Raised for invalid requests (HTTP 400). */
export class BadRequestError extends KiBrowserError {
  constructor(message: string) {
    super(message, 400);
    this.name = "BadRequestError";
  }
}

// ============================================================================
// API Response Wrapper
// ============================================================================

interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: string;
}

// ============================================================================
// SDK Client
// ============================================================================

/**
 * TypeScript SDK client for the ki-browser REST API.
 *
 * Provides async methods for all browser automation endpoints including
 * tab management, navigation, DOM interaction, and screenshots.
 *
 * @example
 * ```typescript
 * const browser = new KiBrowser("http://localhost:9222");
 *
 * // Check health
 * const health = await browser.health();
 * console.log(`Server: ${health.version}`);
 *
 * // Navigate and take a screenshot
 * await browser.navigate("https://example.com");
 * const shot = await browser.screenshot({ format: "png" });
 * console.log(`Screenshot: ${shot.width}x${shot.height}`);
 * ```
 */
export class KiBrowser {
  private baseUrl: string;
  private timeout: number;

  /**
   * Create a new KiBrowser client.
   *
   * @param baseUrl - Base URL of the ki-browser API server.
   * @param timeout - Request timeout in milliseconds. Defaults to 30000.
   */
  constructor(baseUrl: string = "http://localhost:9222", timeout: number = 30000) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.timeout = timeout;
  }

  // ========================================================================
  // Internal helpers
  // ========================================================================

  private url(path: string): string {
    return `${this.baseUrl}${path}`;
  }

  private async handleResponse<T>(resp: Response): Promise<T> {
    if (resp.status === 503) {
      throw new ApiDisabledError();
    }

    const body: ApiResponse<T> = await resp.json();

    if (!body.success) {
      const msg = body.error ?? "Unknown error";
      if (resp.status === 400) {
        throw new BadRequestError(msg);
      }
      throw new KiBrowserError(msg, resp.status);
    }

    return body.data as T;
  }

  private async get<T>(path: string, params?: Record<string, string>): Promise<T> {
    const queryString = params
      ? "?" + new URLSearchParams(params).toString()
      : "";

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const resp = await fetch(this.url(path) + queryString, {
        method: "GET",
        headers: { "Content-Type": "application/json" },
        signal: controller.signal,
      });
      return this.handleResponse<T>(resp);
    } finally {
      clearTimeout(timer);
    }
  }

  private async post<T>(path: string, body?: Record<string, unknown>): Promise<T> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const resp = await fetch(this.url(path), {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body ?? {}),
        signal: controller.signal,
      });
      return this.handleResponse<T>(resp);
    } finally {
      clearTimeout(timer);
    }
  }

  // ========================================================================
  // Health
  // ========================================================================

  /** Check server health. */
  async health(): Promise<HealthStatus> {
    return this.get<HealthStatus>("/health");
  }

  // ========================================================================
  // Tab Management
  // ========================================================================

  /** List all open browser tabs. */
  async listTabs(): Promise<TabsResponse> {
    return this.get<TabsResponse>("/tabs");
  }

  /**
   * Create a new browser tab.
   *
   * @param options - Optional URL and active state.
   */
  async newTab(options?: NewTabOptions): Promise<NewTabResponse> {
    return this.post<NewTabResponse>("/tabs/new", options);
  }

  /**
   * Close a browser tab.
   *
   * @param tabId - ID of the tab to close.
   */
  async closeTab(tabId: string): Promise<void> {
    await this.post<void>("/tabs/close", { tab_id: tabId });
  }

  // ========================================================================
  // Navigation & Interaction
  // ========================================================================

  /**
   * Navigate to a URL.
   *
   * @param url - The URL to navigate to.
   * @param tabId - Target tab ID. Uses active tab if omitted.
   */
  async navigate(url: string, tabId?: string): Promise<void> {
    const body: Record<string, unknown> = { url };
    if (tabId !== undefined) body.tab_id = tabId;
    await this.post<void>("/navigate", body);
  }

  /**
   * Click at coordinates or on a CSS selector.
   *
   * Provide either (x, y) coordinates or a CSS selector.
   */
  async click(options: ClickOptions): Promise<void> {
    await this.post<void>("/click", options as Record<string, unknown>);
  }

  /**
   * Type text into the focused element or a specified selector.
   *
   * @param text - Text to type.
   * @param options - Optional selector, clear_first, and tab_id.
   */
  async type(text: string, options?: TypeOptions): Promise<void> {
    await this.post<void>("/type", { text, ...options });
  }

  /**
   * Scroll the page.
   *
   * @param options - Scroll position, delta, selector, behavior, tab_id.
   */
  async scroll(options: ScrollOptions): Promise<void> {
    await this.post<void>("/scroll", options as Record<string, unknown>);
  }

  /**
   * Execute JavaScript in the browser context.
   *
   * @param script - JavaScript code to evaluate.
   * @param options - Optional await_promise and tab_id.
   * @returns The evaluation result.
   */
  async evaluate(script: string, options?: EvaluateOptions): Promise<unknown> {
    const data = await this.post<EvaluateResponse>("/evaluate", {
      script,
      ...options,
    });
    return data.result;
  }

  // ========================================================================
  // Screenshots
  // ========================================================================

  /**
   * Capture a screenshot of the current page.
   *
   * @param options - Format, quality, full_page, selector, tab_id.
   */
  async screenshot(options?: ScreenshotOptions): Promise<ScreenshotResponse> {
    const params: Record<string, string> = {};
    if (options?.format) params.format = options.format;
    if (options?.quality !== undefined) params.quality = String(options.quality);
    if (options?.full_page) params.full_page = "true";
    if (options?.selector) params.selector = options.selector;
    if (options?.tab_id) params.tab_id = options.tab_id;

    return this.get<ScreenshotResponse>("/screenshot", params);
  }

  // ========================================================================
  // DOM Operations
  // ========================================================================

  /**
   * Find a DOM element by CSS selector.
   *
   * @param selector - CSS selector to search for.
   * @param options - Optional timeout and tab_id.
   */
  async findElement(selector: string, options?: FindElementOptions): Promise<ElementInfo> {
    const params: Record<string, string> = { selector };
    if (options?.timeout !== undefined) params.timeout = String(options.timeout);
    if (options?.tab_id) params.tab_id = options.tab_id;

    return this.get<ElementInfo>("/dom/element", params);
  }

  // ========================================================================
  // API Management
  // ========================================================================

  /** Get current API server status. */
  async apiStatus(): Promise<ApiStatus> {
    return this.get<ApiStatus>("/api/status");
  }

  /**
   * Toggle API enabled state.
   *
   * @param enabled - Whether to enable or disable the API.
   */
  async toggleApi(enabled: boolean): Promise<ApiStatus> {
    return this.post<ApiStatus>("/api/toggle", { enabled });
  }
}

// ============================================================================
// Example Usage
// ============================================================================

async function main() {
  const browser = new KiBrowser("http://localhost:9222");

  // Check health
  const health = await browser.health();
  console.log(`Server: ${health.version} - ${health.status}`);

  // Create a tab and navigate
  const tab = await browser.newTab({ url: "https://example.com" });
  console.log(`Created tab: ${tab.tab_id}`);

  // List tabs
  const tabs = await browser.listTabs();
  for (const t of tabs.tabs) {
    console.log(`  Tab ${t.id}: ${t.title} (${t.url})`);
  }

  // Take a screenshot
  const shot = await browser.screenshot({ format: "png" });
  console.log(`Screenshot: ${shot.width}x${shot.height}`);

  // Find an element
  const el = await browser.findElement("h1");
  if (el.found) {
    console.log(`Found <${el.tag_name}>: ${el.text_content}`);
  }

  // Run JavaScript
  const result = await browser.evaluate("document.title");
  console.log(`Page title: ${result}`);

  // Close the tab
  await browser.closeTab(tab.tab_id);
}

// Run example if executed directly
if (typeof require !== "undefined" && require.main === module) {
  main().catch(console.error);
}

File: sdks/python/ki_browser.py
"""
ki-browser Python SDK

Handwritten SDK for the ki-browser REST API.
Provides a clean, typed interface for browser automation via HTTP.

Requirements:
    pip install requests

Usage:
    from ki_browser import KiBrowser

    browser = KiBrowser("http://localhost:9222")
    browser.navigate("https://example.com")
    tabs = browser.list_tabs()
    screenshot = browser.screenshot()
"""

from __future__ import annotations

import base64
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

import requests


# ============================================================================
# Response Types
# ============================================================================


@dataclass
class HealthStatus:
    """Health check response from the API server."""

    status: str
    version: str
    api_enabled: bool


@dataclass
class TabInfo:
    """Information about a browser tab."""

    id: str
    url: str
    title: str
    is_loading: bool
    is_active: bool
    can_go_back: bool
    can_go_forward: bool


@dataclass
class TabsResponse:
    """Response from listing all tabs."""

    tabs: List[TabInfo]
    active_tab_id: Optional[str]


@dataclass
class NewTabResponse:
    """Response from creating a new tab."""

    tab_id: str
    url: str


@dataclass
class BoundingBox:
    """Bounding box of a DOM element in pixels."""

    x: float
    y: float
    width: float
    height: float


@dataclass
class ElementInfo:
    """Information about a found DOM element."""

    found: bool
    tag_name: Optional[str] = None
    text_content: Optional[str] = None
    attributes: Optional[Dict[str, Any]] = None
    bounding_box: Optional[BoundingBox] = None
    is_visible: Optional[bool] = None


@dataclass
class ScreenshotResponse:
    """Screenshot capture response with base64 encoded image data."""

    data: str
    format: str
    width: int
    height: int

    def save(self, path: str) -> None:
        """Save the screenshot to a file."""
        raw = base64.b64decode(self.data)
        with open(path, "wb") as f:
            f.write(raw)


@dataclass
class EvaluateResponse:
    """Response from JavaScript evaluation."""

    result: Any


@dataclass
class ApiStatus:
    """API server status."""

    enabled: bool
    port: int
    connected_clients: int


# ============================================================================
# Exceptions
# ============================================================================


class KiBrowserError(Exception):
    """Base exception for ki-browser SDK errors."""

    def __init__(self, message: str, status_code: Optional[int] = None):
        super().__init__(message)
        self.status_code = status_code


class ApiDisabledError(KiBrowserError):
    """Raised when the API is disabled (HTTP 503)."""

    pass


class BadRequestError(KiBrowserError):
    """Raised for invalid requests (HTTP 400)."""

    pass


# ============================================================================
# SDK Client
# ============================================================================


class KiBrowser:
    """Python SDK client for the ki-browser REST API.

    Provides methods for all browser automation endpoints including
    tab management, navigation, DOM interaction, and screenshots.

    Args:
        base_url: Base URL of the ki-browser API server (e.g. "http://localhost:9222").
        timeout: Request timeout in seconds. Defaults to 30.

    Example:
        >>> browser = KiBrowser("http://localhost:9222")
        >>> browser.navigate("https://example.com")
        >>> tabs = browser.list_tabs()
        >>> print(tabs.tabs[0].title)
    """

    def __init__(self, base_url: str = "http://localhost:9222", timeout: int = 30):
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()
        self.session.headers.update({"Content-Type": "application/json"})

    def _url(self, path: str) -> str:
        """Build a full URL from a path."""
        return f"{self.base_url}{path}"

    def _handle_response(self, resp: requests.Response) -> Dict[str, Any]:
        """Parse API response and raise on errors."""
        if resp.status_code == 503:
            raise ApiDisabledError("API is disabled", status_code=503)

        data = resp.json()

        if not data.get("success", False):
            error_msg = data.get("error", "Unknown error")
            if resp.status_code == 400:
                raise BadRequestError(error_msg, status_code=400)
            raise KiBrowserError(error_msg, status_code=resp.status_code)

        return data.get("data") or {}

    def _get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Send a GET request."""
        resp = self.session.get(self._url(path), params=params, timeout=self.timeout)
        return self._handle_response(resp)

    def _post(self, path: str, json: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Send a POST request."""
        resp = self.session.post(self._url(path), json=json or {}, timeout=self.timeout)
        return self._handle_response(resp)

    # ========================================================================
    # Health
    # ========================================================================

    def health(self) -> HealthStatus:
        """Check server health.

        Returns:
            HealthStatus with server status, version, and API enabled flag.
        """
        data = self._get("/health")
        return HealthStatus(
            status=data["status"],
            version=data["version"],
            api_enabled=data["api_enabled"],
        )

    # ========================================================================
    # Tab Management
    # ========================================================================

    def list_tabs(self) -> TabsResponse:
        """List all open browser tabs.

        Returns:
            TabsResponse with list of tabs and active tab ID.
        """
        data = self._get("/tabs")
        tabs = [
            TabInfo(
                id=t["id"],
                url=t["url"],
                title=t["title"],
                is_loading=t["is_loading"],
                is_active=t["is_active"],
                can_go_back=t["can_go_back"],
                can_go_forward=t["can_go_forward"],
            )
            for t in data.get("tabs", [])
        ]
        return TabsResponse(tabs=tabs, active_tab_id=data.get("active_tab_id"))

    def new_tab(self, url: Optional[str] = None, active: bool = True) -> NewTabResponse:
        """Create a new browser tab.

        Args:
            url: URL to open in the new tab. Defaults to about:blank.
            active: Whether to make the new tab active. Defaults to True.

        Returns:
            NewTabResponse with the new tab ID and URL.
        """
        body: Dict[str, Any] = {"active": active}
        if url is not None:
            body["url"] = url
        data = self._post("/tabs/new", json=body)
        return NewTabResponse(tab_id=data["tab_id"], url=data["url"])

    def close_tab(self, tab_id: str) -> None:
        """Close a browser tab.

        Args:
            tab_id: ID of the tab to close.
        """
        self._post("/tabs/close", json={"tab_id": tab_id})

    # ========================================================================
    # Navigation & Interaction
    # ========================================================================

    def navigate(self, url: str, tab_id: Optional[str] = None) -> None:
        """Navigate to a URL.

        Args:
            url: The URL to navigate to.
            tab_id: Target tab ID. Uses active tab if omitted.
        """
        body: Dict[str, Any] = {"url": url}
        if tab_id is not None:
            body["tab_id"] = tab_id
        self._post("/navigate", json=body)

    def click(
        self,
        x: Optional[int] = None,
        y: Optional[int] = None,
        selector: Optional[str] = None,
        button: str = "left",
        tab_id: Optional[str] = None,
    ) -> None:
        """Click at coordinates or on a CSS selector.

        Provide either (x, y) coordinates or a CSS selector, not both.

        Args:
            x: X coordinate to click.
            y: Y coordinate to click.
            selector: CSS selector of element to click.
            button: Mouse button ("left", "right", "middle"). Defaults to "left".
            tab_id: Target tab ID. Uses active tab if omitted.
        """
        body: Dict[str, Any] = {"button": button}
        if x is not None:
            body["x"] = x
        if y is not None:
            body["y"] = y
        if selector is not None:
            body["selector"] = selector
        if tab_id is not None:
            body["tab_id"] = tab_id
        self._post("/click", json=body)

    def type_text(
        self,
        text: str,
        selector: Optional[str] = None,
        clear_first: bool = False,
        tab_id: Optional[str] = None,
    ) -> None:
        """Type text into the focused element or a specified selector.

        Args:
            text: Text to type.
            selector: CSS selector of element to type into.
            clear_first: Whether to clear the field before typing.
            tab_id: Target tab ID. Uses active tab if omitted.
        """
        body: Dict[str, Any] = {"text": text, "clear_first": clear_first}
        if selector is not None:
            body["selector"] = selector
        if tab_id is not None:
            body["tab_id"] = tab_id
        self._post("/type", json=body)

    def scroll(
        self,
        x: Optional[int] = None,
        y: Optional[int] = None,
        delta_x: Optional[int] = None,
        delta_y: Optional[int] = None,
        selector: Optional[str] = None,
        behavior: Optional[str] = None,
        tab_id: Optional[str] = None,
    ) -> None:
        """Scroll the page.

        Args:
            x: Absolute scroll X position.
            y: Absolute scroll Y position.
            delta_x: Relative horizontal scroll amount.
            delta_y: Relative vertical scroll amount.
            selector: CSS selector of element to scroll into view.
            behavior: Scroll behavior ("auto", "smooth", "instant").
            tab_id: Target tab ID. Uses active tab if omitted.
        """
        body: Dict[str, Any] = {}
        if x is not None:
            body["x"] = x
        if y is not None:
            body["y"] = y
        if delta_x is not None:
            body["delta_x"] = delta_x
        if delta_y is not None:
            body["delta_y"] = delta_y
        if selector is not None:
            body["selector"] = selector
        if behavior is not None:
            body["behavior"] = behavior
        if tab_id is not None:
            body["tab_id"] = tab_id
        self._post("/scroll", json=body)

    def evaluate(
        self,
        script: str,
        await_promise: bool = True,
        tab_id: Optional[str] = None,
    ) -> Any:
        """Execute JavaScript in the browser context.

        Args:
            script: JavaScript code to evaluate.
            await_promise: Whether to await the result if it is a Promise.
            tab_id: Target tab ID. Uses active tab if omitted.

        Returns:
            The evaluation result (parsed from JSON).
        """
        body: Dict[str, Any] = {"script": script, "await_promise": await_promise}
        if tab_id is not None:
            body["tab_id"] = tab_id
        data = self._post("/evaluate", json=body)
        return data.get("result")

    # ========================================================================
    # Screenshots
    # ========================================================================

    def screenshot(
        self,
        format: str = "png",
        quality: Optional[int] = None,
        full_page: bool = False,
        selector: Optional[str] = None,
        tab_id: Optional[str] = None,
    ) -> ScreenshotResponse:
        """Capture a screenshot of the current page.

        Args:
            format: Image format ("png" or "jpeg"). Defaults to "png".
            quality: JPEG quality (0-100). Only used for JPEG format.
            full_page: Whether to capture the full scrollable page.
            selector: CSS selector of element to capture.
            tab_id: Target tab ID. Uses active tab if omitted.

        Returns:
            ScreenshotResponse with base64-encoded image data.
        """
        params: Dict[str, Any] = {"format": format}
        if quality is not None:
            params["quality"] = quality
        if full_page:
            params["full_page"] = "true"
        if selector is not None:
            params["selector"] = selector
        if tab_id is not None:
            params["tab_id"] = tab_id
        data = self._get("/screenshot", params=params)
        return ScreenshotResponse(
            data=data["data"],
            format=data["format"],
            width=data["width"],
            height=data["height"],
        )

    # ========================================================================
    # DOM Operations
    # ========================================================================

    def find_element(
        self,
        selector: str,
        timeout: Optional[int] = None,
        tab_id: Optional[str] = None,
    ) -> ElementInfo:
        """Find a DOM element by CSS selector.

        Args:
            selector: CSS selector to search for.
            timeout: Timeout in milliseconds to wait for the element.
            tab_id: Target tab ID. Uses active tab if omitted.

        Returns:
            ElementInfo with element details if found.
        """
        params: Dict[str, Any] = {"selector": selector}
        if timeout is not None:
            params["timeout"] = timeout
        if tab_id is not None:
            params["tab_id"] = tab_id
        data = self._get("/dom/element", params=params)
        bbox = None
        if data.get("bounding_box"):
            b = data["bounding_box"]
            bbox = BoundingBox(x=b["x"], y=b["y"], width=b["width"], height=b["height"])
        return ElementInfo(
            found=data["found"],
            tag_name=data.get("tag_name"),
            text_content=data.get("text_content"),
            attributes=data.get("attributes"),
            bounding_box=bbox,
            is_visible=data.get("is_visible"),
        )

    # ========================================================================
    # API Management
    # ========================================================================

    def api_status(self) -> ApiStatus:
        """Get current API server status.

        Returns:
            ApiStatus with enabled flag, port, and connected client count.
        """
        data = self._get("/api/status")
        return ApiStatus(
            enabled=data["enabled"],
            port=data["port"],
            connected_clients=data["connected_clients"],
        )

    def toggle_api(self, enabled: bool) -> ApiStatus:
        """Toggle API enabled state.

        Args:
            enabled: Whether to enable or disable the API.

        Returns:
            Updated ApiStatus.
        """
        data = self._post("/api/toggle", json={"enabled": enabled})
        return ApiStatus(
            enabled=data["enabled"],
            port=data["port"],
            connected_clients=data["connected_clients"],
        )


# ============================================================================
# Example Usage
# ============================================================================

if __name__ == "__main__":
    browser = KiBrowser("http://localhost:9222")

    # Check health
    health = browser.health()
    print(f"Server: {health.version} - {health.status}")

    # Create a tab and navigate
    tab = browser.new_tab("https://example.com")
    print(f"Created tab: {tab.tab_id}")

    # List tabs
    tabs = browser.list_tabs()
    for t in tabs.tabs:
        print(f"  Tab {t.id}: {t.title} ({t.url})")

    # Take a screenshot
    shot = browser.screenshot(format="png")
    shot.save("screenshot.png")
    print(f"Screenshot saved: {shot.width}x{shot.height}")

    # Find an element
    el = browser.find_element("h1")
    if el.found:
        print(f"Found <{el.tag_name}>: {el.text_content}")

    # Run JavaScript
    result = browser.evaluate("document.title")
    print(f"Page title: {result}")

    # Close the tab
    browser.close_tab(tab.tab_id)
